### PR TITLE
Replace ::stat call with std::filesystem equivalents

### PR DESCRIPTION
**Pull request by JoeLoser**
###  :notebook_with_decorative_cover: Description

Problem:
- Use of `::stat` system call is discouraged when it can be replaced
  with higher level standard facilities from `std::filesystem`.

Solution:
- Replace `::stat` call used to get the last modified time and file size
  directly with their equivalent functions from `std::filesystem`. This
  also allows us to work with strong types in the `partition_diskstate`
  which is nice.

###  :memo: Checklist

- [X] All user-facing changes have changelog entries.
- [X] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [X] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

File-by-file.

### CODE REVIEW

**Comment by JoeLoser** [comment]
@mavam I see one of the integration tests is failing. Can you help me understand why as a result of this change?

**Review by tobim — DISMISSED** [review]
Thank you for trying to improve this code. I'm not sure that a conversion to use the `std::filesystem` API actually achieves that though. 

The filesystem functions must be implemented in terms of syscalls as well, and I'm pretty sure they will both just call `stat` internally. Calling `stat` twice as often as necessary needs a good justification.

Is your goal to port the code to a non-posix platform?

File: libvast/src/system/disk_monitor.cpp
Comment (JoeLoser) [reply libvast/src/system/disk_monitor.cpp:18]: > Thank you for trying to improve this code. I'm not sure that a conversion to use the `std::filesystem` API actually achieves that though.
> 
> The filesystem functions must be implemented in terms of syscalls as well, and I'm pretty sure they will both just call `stat` internally. Calling `stat` twice as often as necessary needs a good justification.
> 
> Is your goal to port the code to a non-posix platform?

I don't have a goal of porting in mind, but not making system calls directly would help in that regard. This was just a drive-by refactoring when reading the code as I noticed the other day that all of `vast::path` and `vast::directory` could be replaced with their `std::filesystem` equivalents. Personally (others may not share this same opinion), I find the `std::filesystem` functions easier to read than C-style code with making a struct that gets filled in by a system call. This is more canonical C++ I'd argue.

If performance is a concern here in the disk monitor, I'm fine closing the MR as this change would result in twice as many `stat` system calls as you pointed out I suspect.

File: libvast/src/system/disk_monitor.cpp
Comment (JoeLoser) [reply libvast/src/system/disk_monitor.cpp:94]: I can add that, no problem. Let's decide if we want to keep pursuing this MR at all before I make any modifications to it.

File: libvast/src/system/disk_monitor.cpp
Comment (lava) [reply libvast/src/system/disk_monitor.cpp:18]: On the one hand, performance in the "erase" path is probably not a major concern, since this should only be called relatively infrequently.

On the other hand, doing unnecessary work also does not feel  very C++-like.

From how I understand `std::filesystem`, the intended solution here would probably be to use a directory iterator, since struct `directory_entry` already has member functions `file_size` and `last_write_time`, which are allowed to cache the information. Although as far as I can tell only libcxx currently supports the caching.

Comment (JoeLoser) [reply libvast/src/system/disk_monitor.cpp:18]: > On the one hand, performance in the "erase" path is probably not a major concern, since this should only be called relatively infrequently.
> 
> On the other hand, doing unnecessary work also does not feel very C++-like.
> 
> From how I understand `std::filesystem`, the intended solution here would probably be to use a directory iterator, since struct `directory_entry` already has member functions `file_size` and `last_write_time`, which are allowed to cache the information. Although as far as I can tell only libcxx currently supports the caching.

Good observation regarding the the `directory_entry` caching of file status -- I noticed that the other day after this PR opened, but did not yet mention it here. 

Note that https://github.com/tenzir/vast/pull/1411 changes the surrounding structure here to use `std::filesystem` with a `directory_iterator` such that we are working with `directory_entry` types in the loop. This would give rise in this PR to take advantage of that like you mentioned.

I'd say it's a quality of implementation concern with whether `libc++` or `libstdc++` caches the `directory_entry` status info for paths.  I'm glad that you brought up the erase performance here in the `disk_monitor` -- thanks for providing some context.

With that being said, @lava @dominiklohmann, do we want to move forward with this PR after https://github.com/tenzir/vast/pull/1411 lands?

File: libvast/src/system/disk_monitor.cpp
Comment (JoeLoser) [reply libvast/src/system/disk_monitor.cpp:96]: Just added a warning 👍

File: libvast/src/system/disk_monitor.cpp
Comment (dominiklohmann) [reply libvast/src/system/disk_monitor.cpp:18]: I'll leave the decision up to @lava, as he wrote most of the disk monitor and should make this decision.

Comment (mavam) [reply libvast/src/system/disk_monitor.cpp:18]: I agree that @lava should pull the trigger here.

My 2 cents: assume that the caching takes place and work with `directory_entry`.

Comment (lava) [reply libvast/src/system/disk_monitor.cpp:18]: Yes, I agree, I'm fine with merging this after a rebase.

Comment (JoeLoser) [reply libvast/src/system/disk_monitor.cpp:17]: @lava just rebased. Please take a look again. Thanks!

File: libvast/src/system/disk_monitor.cpp
Comment (lava) [inline libvast/src/system/disk_monitor.cpp:113]: Didn't you want to switch this to `file.file_size(ec)`/`file.last_write_time(ec)` after #1411 has been merged?

Comment (JoeLoser) [reply libvast/src/system/disk_monitor.cpp:113]: Yes, thanks for the reminder. I will make that change later tonight -- sorry about that.

Comment (JoeLoser) [reply libvast/src/system/disk_monitor.cpp]: CI is passing now with using the `file_size` and `last_write_time` directly as member function calls from the corresponding `directory_entry` type as discussed.

File: libvast/src/system/disk_monitor.cpp
Comment (mavam) [inline libvast/src/system/disk_monitor.cpp:111]:
```suggestion
```

File: libvast/src/system/disk_monitor.cpp
Comment (mavam) [inline libvast/src/system/disk_monitor.cpp:116]: I [tried to find](https://en.cppreference.com/w/cpp/filesystem/directory_entry/last_write_time) when `last_write_time` can return a value with all bits set to 1, and why it's an error. At least high-level docs don't hint at an issue. Mind sharing you insight on this check?

File: libvast/src/system/disk_monitor.cpp
Comment (mavam) [inline libvast/src/system/disk_monitor.cpp:117]:
```suggestion
            partitions.emplace_back(id, file_size, mtime);
```

**Review by mavam — APPROVED** [review]
Looks good, just a few nits.

File: libvast/src/system/disk_monitor.cpp
Comment (JoeLoser) [reply libvast/src/system/disk_monitor.cpp:116]: The first part of the `if` check here is to catch the error case for `last_write_time` per https://en.cppreference.com/w/cpp/filesystem/directory_entry/last_write_time where it states:

> The overload taking a std::error_code& parameter sets it to the OS API error code if an OS API call fails

The second part of the `if` check here is to catch the `file_size` error case per https://en.cppreference.com/w/cpp/filesystem/file_size where it states:

> The non-throwing overload returns static_cast<std::uintmax_t>(-1) on errors.

Does that answer your question? I think you may have misread the code as the check isn't for `last_write_time` having all its bits set.

File: libvast/src/system/disk_monitor.cpp
Comment (JoeLoser) [reply libvast/src/system/disk_monitor.cpp]: Just changed to use `emplace_back` so we get the move construction here. 👍

**Comment by JoeLoser** [comment]
> Looks good, just a few nits.

Thanks, just pushed up the corresponding changes. Let me know if there is anything else. Appreciate the review.

File: libvast/src/system/disk_monitor.cpp
Comment (JoeLoser) [reply libvast/src/system/disk_monitor.cpp]: We can't do an `emplace_back` actually since the `partition_diskstate` isn't move constructible. Just changed it back to use `push_back` which will copy construct the `partition_diskstate` instances.

File: libvast/src/system/disk_monitor.cpp
Comment (JoeLoser) [reply libvast/src/system/disk_monitor.cpp]: ping @lava -- how does this look to you?

File: libvast/src/system/disk_monitor.cpp
Comment (mavam) [reply libvast/src/system/disk_monitor.cpp:115]: Thanks for providing the references.

> I think you may have misread the code as the check isn't for `last_write_time` having all its bits set.

Yes, my knee-jerk reaction to seeing `static_cast<size_t>(-1)` (or for other unsigned arithmetic types) is that someone uses a lazy way to get the largest possible value, instead of using `numeric_limits<size_t>::max()`.

**Comment by mavam** [comment]
Not quite sure what's up with CI. @dominiklohmann?

**Comment by dominiklohmann** [comment]
> Not quite sure what's up with CI. @dominiklohmann?

It's this weird connection issue we're only seeing in CI again. Let's merge regardless.

I've dismissed @tobim's change request since that was addressed already (please do so yourself if you request changes over just commenting), and approved because @lava told me earlier he was just waiting for that change request to disappear. 🤷‍♂️

**Pull request merged**
